src/cargo/ops/cargo_compile.rs

```diff
@@ -534,307 +548,308 @@ impl CompileFilter {
     }
 }
 
-#[derive(Clone, Copy, Debug)]
-struct BuildProposal<'a> {
-    target: &'a Target,
-    profile: &'a Profile,
-    required: bool,
-}
-
-fn generate_default_targets<'a>(
+/// Generates all the base targets for the packages the user has requested to
+/// compile. Dependencies for these targets are computed later in
+/// `unit_dependencies`.
+fn generate_targets<'a>(
+    packages: &[&'a Package],
+    filter: &CompileFilter,
+    default_arch_kind: Kind,
     mode: CompileMode,
-    targets: &'a [Target],
-    profile: &'a Profile,
-    dep: &'a Profile,
-    required_features_filterable: bool,
-) -> Vec<BuildProposal<'a>> {
-    match mode {
-        CompileMode::Bench => targets
-            .iter()
-            .filter(|t| t.benched())
-            .map(|t| BuildProposal {
-                target: t,
-                profile,
-                required: !required_features_filterable,
-            })
-            .collect::<Vec<_>>(),
-        CompileMode::Test => {
-            let mut base = targets
-                .iter()
-                .filter(|t| t.tested())
-                .map(|t| BuildProposal {
-                    target: t,
-                    profile: if t.is_example() { dep } else { profile },
-                    required: !required_features_filterable,
-                })
-                .collect::<Vec<_>>();
-
-            // Always compile the library if we're testing everything as
-            // it'll be needed for doctests
-            if let Some(t) = targets.iter().find(|t| t.is_lib()) {
-                if t.doctested() {
-                    base.push(BuildProposal {
-                        target: t,
-                        profile: dep,
-                        required: !required_features_filterable,
-                    });
+    resolve: &Resolve,
+) -> CargoResult<Vec<Unit<'a>>> {
+    let mut units = Vec::new();
+
+    // Helper for creating a Unit struct.
+    let new_unit =
+        |pkg: &'a Package, target: &'a Target, mode: CompileMode, profile_for: ProfileFor| {
+            let actual_profile_for = if profile_for != ProfileFor::Any {
```

Comment:
Just a personal taste, but I would name this just `profile_for` to shadow the input parameter, so that, down the line, we can't use the wrong `profile_for`.